src/main/cpp/src/regex_rewrite_utils.cu

```diff
@@ -0,0 +1,142 @@
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/strings/detail/utf8.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/find.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda/atomic>
+#include <thrust/binary_search.h>
+#include <thrust/fill.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform.h>
+
+namespace spark_rapids_jni {
+
+/**
+ * @brief Utility to return a bool column indicating the presence of
+ * a given prefix string followed by at leasat range_len characters
+ * with code points in the range [start, end] in each string of the input column.
+ *
+ * @tparam BoolFunction Return bool value given two strings.
+ *
+ * @param strings Column of strings to check for prefix.
+ * @param prefix UTF-8 encoded string to check in strings column.
+ * @param range_len Minimum number of characters to check after the prefix.
+ * @param start Minimum code point value to check for in the range.
+ * @param end Maximum code point value to check for in the range.
+ * @param pfn Returns bool value if prefix is found in the given string.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @param mr Device memory resource used to allocate the returned column's device memory.
+ * @return New BOOL column.
+ */
+template <typename BoolFunction>
+std::unique_ptr<cudf::column> literal_range_pattern_fn(cudf::strings_column_view const& strings,
+                                                       cudf::string_scalar const& prefix,
+                                                       int const range_len,
+                                                       int const start,
+                                                       int const end,
+                                                       BoolFunction pfn,
+                                                       rmm::cuda_stream_view stream,
+                                                       rmm::mr::device_memory_resource* mr)
+{
+  auto strings_count = strings.size();
+  if (strings_count == 0) return cudf::make_empty_column(cudf::type_id::BOOL8);
+
+  CUDF_EXPECTS(prefix.is_valid(stream), "Parameter prefix must be valid.");
+
+  auto d_prefix       = cudf::string_view(prefix.data(), prefix.size());
+  auto strings_column = cudf::column_device_view::create(strings.parent(), stream);
+  auto d_strings      = *strings_column;
```

Comment:
`d_strings` should be a reference to avoid copying.
```suggestion
  auto const d_prefix       = cudf::string_view(prefix.data(), prefix.size());
  auto const strings_column = cudf::column_device_view::create(strings.parent(), stream);
  auto const& d_strings      = *strings_column;
```